torquedata/torquedata/__init__.py

```diff
@@ -5,12 +5,68 @@
 import os
 import pickle
 import json
+from dataclasses import dataclass
+from collections import UserDict
 from whoosh import index
 from whoosh.index import create_in
 from whoosh.fields import *
 
 app = Flask(__name__)
 
+# explanation: since this class inherits from UserDict, it behaves pretty much
+# exactly like a dict which means old code doesn't break but new code can use
+# the methods attached to this class
+class Spreadsheet(UserDict):
+    @classmethod
+    def read(cls, name):
+        """Load an existing spreadsheet"""
+        raise NotImplementedError
+
+    @classmethod
+    def write(cls):
+        """Save spreadsheet to filesystem"""
+        raise NotImplementedError
+
+    def __init__(self, name, object_name, key_column, columns=[]):
+        self.name = name
+        self.object_name = object_name
+        self.key_column = key_column
+        self.columns = columns
+
+        super().__init__()
+
+    def apply_permissions(self, permissions):
+        """
+        Return a new spreadsheet with only data that is allowed by a given
+        permissions object
+        """
+        new_data = {}
+        # cull invalid rows, columns
+        for key, entry in self.data.items():
+            if 'valid_ids' in permissions and key not in permissions['valid_ids']:
+                continue
+            if 'columns' in permissions:
+                new_entry = {}
+                for column, value in entry.items():
+                    if column not in permissions['columns']:
+                        continue
+                    new_entry[column] = value
+                new_data[key] = new_entry
+            else:
+                new_data[key] = entry
+
+        new_sheet = Spreadsheet(
+            name=self.name,
+            object_name=self.object_name,
+            key_column=self.key_column,
+            columns=permissions.get('columns', None) or self.columns
+        )
+        new_sheet.data = new_data
+        return new_sheet
+
+    def json(self):
+        return json.dumps({self.name: self.data.values()})
+
```

Comment:
This is out of the scope of this issue, but I started adding an interface for spreadsheets that I think is nicer to work with. All the other code works with or without it except for the new API endpoint. Is this something I should keep pursuing?

Reply:
Well, there are two separate questions here, I think:

1) Is it okay to introduce this new class and use it in the new code that is introduced in this commit?

2) Do we want to go back and modify existing code to take advantage of this new class?  (If so, doing that would be a separate change, out of scope for this PR and this issue.)

I'm inclined to say "yes" on (1) and "maybe" on (2), but I'd like to know what @frankduncan thinks.